api/models/datastore.go

```diff
@@ -59,6 +59,24 @@ type Datastore interface {
 	// ErrDatastoreEmptyRoutePath when routePath is empty. Returns ErrRoutesNotFound when no route exists.
 	RemoveRoute(ctx context.Context, appID, routePath string) error
 
+	// PutFn inserts a new function if one does not exist, applying any defaults necessary, or
+	// updates a function that exists under the same name. Returns ErrDatastoreEmptyFn if func is nil,
+	// ErrDatastoreEmptyFnName is func.Name is empty.
+	// TODO(reed): should we allow rename if id provided?
+	PutFn(ctx context.Context, fn *Fn) (*Fn, error)
+
+	// GetFns returns a list of funcs, applying any additional filters provided.
+	GetFns(ctx context.Context, filter *FnFilter) ([]*Fn, error)
+
+	// GetFn returns a function by name. Returns ErrDatastoreEmptyFnName if funcName is empty.
+	// Returns ErrFnsNotFound if a func is not found.
+	// TODO(reed): figure out addressable by id or name biz. iff 1 query, name works.
+	GetFn(ctx context.Context, funcName string) (*Fn, error)
+
+	// RemoveFn removes a function. Returns ErrDatastoreEmptyFnName if funcName is empty.
+	// Returns ErrFnsNotFound if a func is not found.
+	RemoveFn(ctx context.Context, funcName string) error
```

Comment:
these all (except PutFn which has it in the model) need an arg for `appID` I think (like Route) -- maybe we could get away with by name but we never quite sorted that out so I'm not sure, appID is the safe route.

Reply:
I think at this level it should use FnID, the mapping over from the name, appId pair should happen higher up.